### ISSUE TITLE
avformat_producer: demuxing and decoding should be pipelined

### ISSUE THREAD

**Issue by Tjoppen** [mltframework/mlt#870]
Currently demuxing and decoding is done in the same thread in producer_avformat.c. This prevents playing files that have very high bitrate, since the demuxer cannot work while the decoder is doings its thing. The samples at https://dpel.aswf.io/ are examples of files that trigger this problem, for example ASC_StEM2_178_UHD_ST2084_1000nits_Rec2020_Stereo_ProRes4444XQ.mov which is just above 1 Gbit/s. Disk access thus cuts into decoding time, preventing realtime playback in many cases where it would be otherwise possible. On one machine ffmpeg is able to decode this file at ~1.4x realtime whereas melt only manages 0.95x realtime.

I gave real_time=2 on the consumer a try but this didn't improve the situation, less so even higher values. buffer=10 also didn't help. Perhaps I'm missing something. Simplest command that displays this:

```bash
$ time melt -producer avformat:$HOME/samples/dpel.aswf.io/ASC_StEM2_178_UHD_ST2084_1000nits_Rec2020_Stereo_ProRes4444XQ.mov in=1 out=2400 -consumer avformat:/dev/null f=null real_time=2
[snip]
real    1m45.118s
user    13m8.348s
sys     1m35.049s
```

Obviously hard to test without access to the same machine but the general idea should be clear: it takes more than 100s to decode 100s of essence that ffmpeg manages to decode in 70s. For reference just reading 100s from disk with dd takes 45s (bs=10M).

**Comment by Tjoppen** [comment]
I've started implementing this using pthreads and conditions. I'll send a PR if it pans out. I also discovered that the producer doesn't make use of AVDISCARD_ALL where appropriate.

**Comment by ddennedy** [comment]
fixed in 51e0563ec4af199be12f68d3b32700912df8dac7
